docs(apisix): wrote transformation/transcode-http-to-grpc (#141)
api7/docs: apisix_versioned_docs/version-3.2.0/how-to-guide/transformation/grpc.md

---
title: Transcode HTTP to gRPC
slug: /how-to-guide/transformation/grpc
---

[gRPC](https://grpc.io/) is an open-source high performance Remote Procedure Call (RPC) framework based on HTTP/2 protocol. It uses [protocol buffers (protobuf)](https://protobuf.dev/) as the interface description language (IDL). gRPC framework provides a portable mechanism for load balancing, link tracing, health checking, and authentication, effectively connecting multiple services between datacenters.

> APISIX supports transforming between HTTP RESTful and gRPC using the plugin `grpc-transcode`. The plugin supports specifying the contents of `.proto` files in Apache APISIX and implementing proxies for different gRPC services through user-defined gRPC services.

Reviewer: ```suggestion
APISIX provides the capability to transform between RESTful HTTP and gRPC requests and responses, using the plugin `grpc-transcode` and proto objects. 
 
 [//]: <TODO: hyperlink proto>
```

This guide will show you how to use the plugin `grpc-transcode` to transform RESTful requests to gRPC requests.

<br/>
  <div style={{textAlign: 'center'}}>
    <img src="https://static.apiseven.com/uploads/2023/05/05/Lfadgl5V_transformation-grpc.jpg" alt="REST to gRPC Diagram" width="70%"/>
  </div>
<br/>

## Prerequisite(s)

* Install [Docker](https://docs.docker.com/get-docker/).
* Install [cURL](https://curl.se/) to send requests to APISIX for validation.
* Install [gRPCurl](https://github.com/fullstorydev/grpcurl) to send requests to gRPC services for validation.
* Follow the [Getting Started tutorial](./../../getting-started/) to start a new APISIX instance in Docker.

> ## Deploy Example gRPC Server

Reviewer: how about
```suggestion
## Deploy a gRPC Server
```

> Spin up an [example gRPC server](https://github.com/api7/grpc_server_example) instance named `quickstart-grpc-example` in Docker, exposed on port `50051`:

Reviewer: ```suggestion
Start an [example gRPC server](https://github.com/api7/grpc_server_example) Docker instance `quickstart-grpc-example` on port `50051`:
```

```shell
docker run -d \
# highlight-start
  --name quickstart-grpc-example \
  --network=apisix-quickstart-net \
  -p 50051:50051 api7/grpc-server-example:1.0.2
# highlight-end
```

:::info

The example gRPC server instance needs to join the network `apisix-quickstart-net` to be accessed by containerized APISIX. Adjust the command arguments accordingly if you use other installation methods for APISIX to ensure that the network works.

:::

This example gRPC server holds several services, such as `echo.EchoService` defined in `echo.proto`:

```proto title="echo.proto"
syntax = "proto3";

package echo;

service EchoService {
  rpc Echo (EchoMsg) returns (EchoMsg);
}

message EchoMsg {
  string msg = 1;
}
```

> For example, the gRPC method `echo.EchoService.Echo` accept parameters of `EchoMsg` type, and respond with `EchoMsg` type also.

Reviewer: ```suggestion
In this example, `Echo` is a method in `EchoService` that accepts parameter type `string` defined in `EchoMsg`.
```

:::

This example gRPC server holds several services, such as `echo.EchoService`:

```proto title="echo.proto"
syntax = "proto3";

package echo;

service EchoService {
  rpc Echo (EchoMsg) returns (EchoMsg);
}

message EchoMsg {
  string msg = 1;
}
```

In this example, `Echo` is a method in `EchoService` that accepts parameter type `string` defined in `EchoMsg`.

Test the gRPC method `echo.EchoService.Echo` using gRPCurl:

```shell
grpcurl -plaintext -d '{"msg": "Hello"}' "127.0.0.1:50051" "echo.EchoService/Echo"
```

A response similar to the following verifies that the gRPC service is working:

```text
{
  "msg": "Hello"
}
```

## Create a Proto Object to Store Protobuf File

Store the protobuf file `echo.proto` as a proto object in APISIX with the id `quickstart-proto`:

```shell
curl -i "http://127.0.0.1:9180/apisix/admin/protos" -X PUT -d '
{
  "id": "quickstart-proto",
  "content": "syntax = \"proto3\";

  package echo;

  service EchoService {
    rpc Echo (EchoMsg) returns (EchoMsg);
  }

  message EchoMsg {
    string msg = 1;
  }"
}'
```

An `HTTP/1.1 201 OK` response verifies that the proto object is created successfully.

## Enable `grpc-transcode` Plugin

> Create a route with id `quickstart-grpc` and enable the plugin `grpc-transcode`. It forwards HTTP requests sent to `/echo` to the upstream gRPC server `quickstart-grpc-example` spun up before:

Reviewer: ```suggestion
Create a route with id `quickstart-grpc` and enable the plugin `grpc-transcode`:
```

```shell
curl -i "http://127.0.0.1:9180/apisix/admin/routes" -X PUT -d '
{
  "id": "quickstart-grpc",
  "methods": ["GET"],
  "uri": "/echo",
  "plugins": {
# highlight-start
    "grpc-transcode": {
// Annotate 1
      "proto_id": "quickstart-proto",
// Annotate 2
      "service": "echo.EchoService",
// Annotate 3
      "method": "Echo"
    }
# highlight-end
  },
  "upstream": {
    "scheme": "grpc",
    "type": "roundrobin",
    "nodes": {
      "quickstart-grpc-example:50051": 1
    }
  }
}'
```

❶ `proto_id`: the proto object which defines gRPC services

❷ `service`: gRPC service `echo.EchoService` in use

❸ `method`: gRPC method `Echo` in use

> The route transforms HTTP Restful requests to gRPC requests and sends them to gRPC server's method `echo.EchoService/Echo`, then transforms the gRPC responses from gRPC server to HTTP Restful responses and sent them to clients.

Reviewer: ```suggestion
APISIX now transcodes the RESTful HTTP requests received at `/echo` route to proto requests and forwards them to the upstream gRPC server `quickstart-grpc-example` to invoke the method `echo.EchoService/Echo`. Once the gRPC server responds, APISIX transcodes the proto responses back to RESTful HTTP responses for clients.
```

❸ `method`: gRPC method `SayHello` in use

The route transforms HTTP Restful requests to gRPC requests and sends them to gRPC server's method `hello.HelloService/SayHello`, then transforms the gRPC responses from gRPC server to HTTP Restful responses and sent them to clients.

An `HTTP/1.1 201 OK` response verifies that the route is created and the plugin `grpc-transcode` is enabled successfully.

## Test gRPC Services in a RESTful Way

Send an HTTP request to `/hello` with parameters defined in `HelloRequest`:

```shell
curl -i http://127.0.0.1:9080/hello?greeting=apisix
```

A valid response similar to the following verifies that the plugin `grpc-transcode` works:

```text
{"reply":"hello apisix"}
```

## Disable `grpc-transcode` Plugin

Disable `grpc-transcode` plugin by setting the `_meta.disable` parameter to `true`.

```shell
curl -i "http://127.0.0.1:9180/apisix/admin/routes/quickstart-grpc" -X PATCH -d '
{
  "plugins": {
    "grpc-transcode": {
      "_meta": {
        "disable": true
      }
    }
  }
}'
```

An `HTTP/1.1 502 Bad Gateway` response verifies that the plugin `grpc-transcode` is disabled successfully.

## Next Steps

> APISIX also supports process gRPC Web requests from JavaScript clients to a gRPC service directly by another plugin `grpc-web`.

Reviewer: ```suggestion

Learn more about the `grpc-transcode` plugin in the plugin reference (coming soon).

In addition to transcoding HTTP requests to gRPC requests, APISIX also supports [gRPC-Web](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-WEB.md), a variation of the [native gRPC protocol](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md), with the APISIX plugin `grpc-web`. 

[//]: <TODO: Add the link to how-to guide or plugin's doc>
```

> [//]: <TODO: grpc-web plugin>

Reviewer: ```suggestion
[//]: <TODO: grpc-transcode plugin reference doc>
[//]: <TODO: grpc-web plugin guide>
```
